File: Data Structure/Trie/LET_no212_Word Search II.cpp
class Trie {
public:
    bool is_word;
    unordered_map<char, Trie*> children;
    
    Trie() { is_word = false; }
    void insert(string& word) {
        Trie* cur = this;
        for (char c : word) {
            if (!cur->children.count(c))
                cur->children[c] = new Trie();
            cur = cur->children[c];
        }
        cur->is_word = true;
    }
    bool search(string& word) {
        Trie* cur = this;
        for (char c : word) {
            if (!cur->children.count(c))
                return false;
            cur = cur->children[c];
        }
        return cur->is_word;
    }
};

class Solution {
public:
    vector<string> findWords(vector<vector<char>>& board, vector<string>& words) {
        vector<string> res;
        Trie trie;
        for (string& w : words)
            trie.insert(w);
        Trie* root = &trie;
        for (int i = 0; i < board.size(); i++) {
            for (int j = 0; j < board[0].size(); j++)
                dfs(board, i, j, root, "", res);
        }
        return res;
    }
    
    void dfs(vector<vector<char>>& board, int i, int j, Trie* cur, string s, vector<string>& res) {
        if (cur->is_word) {
            res.push_back(s);
            cur->is_word = false;
        }
        if (!(0 <= i && i < board.size() && 0 <= j && j < board[0].size()))
            return;
        
        char temp = board[i][j];
        if (!(cur->children.count(temp)))
            return;
        cur = cur->children[temp];
        board[i][j] = '*';
        dfs(board, i + 1, j, cur, s + temp, res);
        dfs(board, i - 1, j, cur, s + temp, res);
        dfs(board, i, j + 1, cur, s + temp, res);
        dfs(board, i, j - 1, cur, s + temp, res);
        board[i][j] = temp;
    }
};

File: Data Structure/Trie/LET_no212_Word Search II.py
class Trie:
    def __init__(self, words):
        self.root = {}
        for w in words:
            self.insert(w)
    def insert(self, word):
        cur = self.root
        for c in word:
            if c not in cur:
                cur[c] = {}
            cur = cur[c]
        cur['#'] = word

class Solution:
    def findWords(self, board: List[List[str]], words: List[str]) -> List[str]:
        res = []
        trie = Trie(words)
        for i in range(len(board)):
            for j in range(len(board[0])):
                if board[i][j] in trie.root:
                    self.dfs(board, i, j, trie.root, res)
        return res
    
    def dfs(self, board, i, j, cur, res):
        c = board[i][j]
        nxt = cur[c]
        
        if '#' in nxt:
            res.append(nxt['#'])
            del nxt['#']
            
        board[i][j] = '*'
        for di, dj in [(0, 1), (1, 0), (0, -1), (-1, 0)]:
            i2, j2 = i + di, j + dj
            if not (0 <= i2 < len(board) and 0 <= j2 < len(board[0])):
                continue
            if board[i2][j2] == '*' or board[i2][j2] not in nxt:
                continue
            self.dfs(board, i2, j2, nxt, res)
        board[i][j] = c
        
        if not cur[c]:
            del cur[c]